### ISSUE TITLE
Proper Browserify Handling

### ISSUE THREAD

**Issue by iamJoeTaylor** [square/field-kit#29]
Right now our `dist` package has a conflict with `require` in browserify.

**Comment by alexgorbatchev** [comment]
I believe that `require` in browserify is scoped to the package and not leaked anywhere (so I don't understand what you mean). Could you add some more details please?

**Comment by eventualbuddha** [comment]
The issue is that if you try to use the browserified field-kit.js file as
input to browserify then browserify will see the `require` calls and think
it needs to go find those files. We should add reproduction steps here…

**Comment by iamJoeTaylor** [comment]
If you start a new project and npm install fieldkit and try to use the
package in a browserify project it will not build.

**Comment by alexgorbatchev** [comment]
Did I understand this correctly - `dist` files can't be used in a CommonJS project and calling `require('./dist/field-kit')` blows up?

**Comment by eventualbuddha** [comment]
```
$ npm i field-kit
$ echo 'require("field-kit")' > index.js
$ browserify -e index.js
Error: module "./amex_card_formatter" not found from "/Users/donovan/Desktop/node_modules/field-kit/dist/field-kit.js"
    at notFound (/usr/local/lib/node_modules/browserify/index.js:811:15)
    at /usr/local/lib/node_modules/browserify/index.js:761:23
    at /usr/local/lib/node_modules/browserify/node_modules/browser-resolve/index.js:185:24
    at /usr/local/lib/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:36:22
    at load (/usr/local/lib/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:54:43)
    at /usr/local/lib/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:60:22
    at /usr/local/lib/node_modules/browserify/node_modules/browser-resolve/node_modules/resolve/lib/async.js:16:47
    at FSReqWrap.oncomplete (fs.js:95:15)
```

**Comment by alexgorbatchev** [comment]
Thanks for the concrete example. The problem is `"main": "dist/field-kit.js"` in `package.json`. This isn't a browserify issue. The main purpose behind browserify is to package up application files to be executed in browser environment. *I think this is expected and correct behavior when it comes to browserify*. For this example to work we need to change `main` in `package.json` to be `"main": "lib/index.js"`.

Without `lib` folder and correct `main` option the example above won't work in any native environment without additional preprocessing.

**Comment by eventualbuddha** [comment]
Both this and #30 seem to come down to philosophical preferences over how npm packages should be built and distributed. I would argue that since the primary users of this library are going to use it in the browser that we should provide a build that works out of the box in the browser. Distributing `lib/` does not do that.

Building a UMD package and using it as the entry point for the package (what we have now) does accomplish that, and with a fairly simple change (adding derequire) we can make it work for the browserify build use case too. Does this seem reasonable?

**Comment by alexgorbatchev** [comment]
Sounds good.

**Comment by iamJoeTaylor** [comment]
Can we please come to a conclusion for this ticket.

**Comment by alexgorbatchev** [comment]
I think the conclusion is to provide a build that works out of the box using UMD package (as per @eventualbuddha).

**Comment by eventualbuddha** [comment]
Let's go with the simple version for now that uses browserify and derequire.

**Comment by iamJoeTaylor** [comment]
I gave this a go, and the `derequire` package works, but our UMD only adds fieldKit on the window and doesn't return it when `require`ing it.

I tried to fix the UMD but to me it seems that the bowserify FieldKit package and the browserify that is requiring it are STILL clobbering each other. When you `module.exports` in FieldKit you are getting the FieldKit `module` based on its module system. I haven't found a way of actually connecting to and exporting on the "includers" module system. We REALLY need to get a fix for this. IMO even if it's rollup I'm fine with that, but this needs to be resolved. @alexgorbatchev care to give this a shot?
